src/theory/symmetry_detect.cpp

```diff
@@ -0,0 +1,576 @@
+  }
+  else
+  {
+    map<Node, PartitionTrie>::iterator part_it = d_children.begin();
+
+    while (part_it != d_children.end())
+    {
+      getNewPartition(part, part_it->second);
+      ++part_it;
+    }
+  }
+}
+
+void SymmetryDetect::PartitionTrie::getNewPartition(Partition& part,
+                                                    PartitionTrie& pt)
+{
+  if (!pt.d_variables.empty())
+  {
+    vector<Node> vars;
+    vector<Node>::iterator var_it = pt.d_variables.begin();
+    Node fresh_var = NodeManager::currentNM()->mkSkolem(
+        "sym_bk", pt.d_variables[0].getType());
+    Trace("sym-dt")
+        << "[sym-dt] A partition from leaves of the partition trie:{";
+
+    while (var_it != pt.d_variables.end())
+    {
+      Trace("sym-dt") << " " << *var_it;
+      vars.push_back(*var_it);
+      part.d_var_to_subvar[*var_it] = fresh_var;
+      ++var_it;
+    }
+    Trace("sym-dt") << " }" << endl;
+    part.d_subvar_to_vars[fresh_var] = vars;
+  }
+  else
+  {
+    map<Node, PartitionTrie>::iterator part_it = pt.d_children.begin();
+    while (part_it != pt.d_children.end())
+    {
+      getNewPartition(part, part_it->second);
+      ++part_it;
+    }
+  }
+}
+
+void SymmetryDetect::getVariables(vector<Partition>& partitions,
+                                  unordered_set<Node, NodeHashFunction>& vars)
+{
+  vector<Partition>::iterator part_it = partitions.begin();
+
+  while (part_it != partitions.end())
+  {
+    map<Node, vector<Node> >::iterator sub_var_it =
+        (*part_it).d_subvar_to_vars.begin();
+
+    while (sub_var_it != (*part_it).d_subvar_to_vars.end())
+    {
+      vector<Node>::iterator v_it = (sub_var_it->second).begin();
+
+      while (v_it != (sub_var_it->second).end())
+      {
+        vars.insert(*v_it);
+        ++v_it;
+      }
+      ++sub_var_it;
+    }
+    ++part_it;
+  }
+}
+
+void SymmetryDetect::PartitionTrie::addNode(Node v,
+                                            vector<Partition>& partitions)
+{
+  Trace("sym-dt") << "[sym-dt] Add a variable {" << v
+                  << "} to the partition trie..." << endl;
+  vector<Node> subvars;
+  vector<Partition>::iterator part_it = partitions.begin();
+
+  while (part_it != partitions.end())
+  {
+    map<Node, Node>::iterator var_sub_it = (*part_it).d_var_to_subvar.find(v);
+
+    if (var_sub_it != (*part_it).d_var_to_subvar.end())
+    {
+      map<Node, vector<Node> >::iterator sub_vars_it =
+          (*part_it).d_subvar_to_vars.begin();
+
+      while (sub_vars_it != (*part_it).d_subvar_to_vars.end())
+      {
+        if (var_sub_it->second == sub_vars_it->first)
+        {
+          subvars.push_back(var_sub_it->second);
+        }
+        else
+        {
+          subvars.push_back(Node::null());
+        }
+        ++sub_vars_it;
+      }
+    }
+    else
+    {
+      map<Node, vector<Node> >::iterator sub_vars_it =
+          (*part_it).d_subvar_to_vars.begin();
+
+      while (sub_vars_it != (*part_it).d_subvar_to_vars.end())
```

Comment:
I think this can be replaced by `subvars.resize((*part_it).d_subvar_to_var.size());`.

Reply:
Changed

Reply:
Btw, do you know why my request failed the checks? Thanks.

Reply:
I found out that the reason for the failure are the trailing whitespaces in src/Makefile.am. I have submitted #1759 to give a clear error message about this in the future.

Reply:
trailing whitespaces are evil™

Reply:
@PaulMeng Replace with range-for.